trackintel/preprocessing/positionfixes.py

```diff
@@ -60,114 +66,54 @@ def generate_staypoints(positionfixes, method='sliding',
-
-                            # TODO Discussion: Is this last point really a staypoint? As we don't know if the
-                            #      person "moves on" afterwards...
-                            if j == num_pfs - 1:
-                                staypoint = {}
-                                staypoint['user_id'] = pfs[j]['user_id']
-                                staypoint[name_geocol] = Point(pfs[j][name_geocol].x, pfs[j][name_geocol].y)
-                                if elevation_flag:
-                                    staypoint['elevation'] = pfs[j]['elevation']
-                                staypoint['started_at'] = pfs[j]['tracked_at']
-                                staypoint['finished_at'] = pfs[j]['tracked_at']
-                                staypoint['id'] = staypoint_id_counter
-
-                                # store matching
-                                posfix_staypoint_matching[staypoint_id_counter] = [
-                                    pfs[j]['id']]  # rather [k for k in range(i, j)]?
-                                staypoint_id_counter += 1
-                                ret_spts = ret_spts.append(staypoint, ignore_index=True)
-                        i = j
-                        break
-                    j = j + 1
-
-            # add matching to original positionfixes (for every user)
-
-            for staypoints_id, posfix_idlist in posfix_staypoint_matching.items():
-                # note that we use .loc because above we have saved the id 
-                # of the positionfixes not thier absolut position
-                ret_pfs.loc[posfix_idlist, 'staypoint_id'] = staypoints_id
-
-
+        ret_spts = ret_pfs.groupby('user_id', as_index=False).apply(_generate_staypoints_sliding_user, 
+                                                                    name_geocol, 
+                                                                    elevation_flag,
+                                                                    dist_threshold, 
+                                                                    time_threshold, 
+                                                                    dist_func).reset_index(drop=True)
+        # spts id management
+        ret_spts['id'] = np.arange(len(ret_spts))
+        ret_spts.set_index('id', inplace=True)
+        
+        # Assign staypoint_id to ret_pfs if spts is detected 
+        if not ret_spts.empty:
+            stps2pfs_map = ret_spts[['pfs_id']].to_dict()['pfs_id']
+            
+            ls = []
+            for key, values in stps2pfs_map.items():
+                for value in values:
+                    ls.append([value, key])
+            temp = pd.DataFrame(ls, columns=['id', 'staypoint_id']).set_index('id')
+            ret_pfs = ret_pfs.join(temp, how='left')
+            ret_spts.drop(columns={'pfs_id'}, inplace=True)
+        # if no staypoint is identified
```

Comment:
This means if no staypoint at all is identified ? Ah, I see the default values create by the join is nan. Good ;-)

Reply:
here 0 staypoint is identified. Otherwise there will be an error in `ret_spts[['pfs_id']].to_dict()['pfs_id']`, because there is no 'pfs_id'